extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/hooks/FiftyOneDeviceDetectionRawAuctionRequestHook.java

```diff
@@ -0,0 +1,672 @@
+package org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.hooks;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.BrandVersion;
+import com.iab.openrtb.request.Device;
+import com.iab.openrtb.request.UserAgent;
+import fiftyone.devicedetection.DeviceDetectionOnPremisePipelineBuilder;
+import fiftyone.devicedetection.DeviceDetectionPipelineBuilder;
+import fiftyone.devicedetection.shared.DeviceData;
+import fiftyone.pipeline.core.data.FlowData;
+import fiftyone.pipeline.core.flowelements.Pipeline;
+import fiftyone.pipeline.engines.Constants;
+import fiftyone.pipeline.engines.data.AspectPropertyValue;
+import fiftyone.pipeline.engines.services.DataUpdateServiceDefault;
+import lombok.Builder;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.MapUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.auction.model.AuctionContext;
+import org.prebid.server.hooks.execution.v1.auction.AuctionRequestPayloadImpl;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.boundary.CollectedEvidence;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.AccountFilter;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.DataFile;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.DataFileUpdate;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.ModuleConfig;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.PerformanceConfig;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.model.ModuleContext;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.model.InvocationResultImpl;
+import org.prebid.server.hooks.v1.InvocationAction;
+import org.prebid.server.hooks.v1.InvocationResult;
+import org.prebid.server.hooks.v1.InvocationStatus;
+import org.prebid.server.hooks.v1.auction.AuctionInvocationContext;
+import org.prebid.server.hooks.v1.auction.AuctionRequestPayload;
+import org.prebid.server.hooks.v1.auction.RawAuctionRequestHook;
+import io.vertx.core.Future;
+import org.prebid.server.proto.openrtb.ext.request.ExtDevice;
+import org.prebid.server.settings.model.Account;
+import org.prebid.server.util.ObjectUtil;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class FiftyOneDeviceDetectionRawAuctionRequestHook implements RawAuctionRequestHook {
```

Comment:
I've split hook into pieces and refactored them a bit. If I messed up the method order it wasn't intentional. The class names might not be ideal as well.

In terms of testing, you can mock the dependencies (I've changed the config for that matter), so you basically don't need protected methods for that.

Not sure whether it makes sense to split classes any further. Maybe the `DeviceEnricher` is still big enough, perhaps additionally extracting the device patching part makes sense.

## Hook

```
public class FiftyOneDeviceDetectionRawAuctionRequestHook implements RawAuctionRequestHook {

    private static final String CODE = "fiftyone-devicedetection-raw-auction-request-hook";

    private final ModuleConfig moduleConfig;
    private final DeviceEnricher deviceEnricher;

    public FiftyOneDeviceDetectionRawAuctionRequestHook(ModuleConfig moduleConfig,
                                                        DeviceEnricher deviceEnricher) {

        this.moduleConfig = Objects.requireNonNull(moduleConfig);
        this.deviceEnricher = Objects.requireNonNull(deviceEnricher);
    }

    @Override
    public String code() {
        return CODE;
    }

    @Override
    public Future<InvocationResult<AuctionRequestPayload>> call(AuctionRequestPayload payload,
                                                                AuctionInvocationContext invocationContext) {

        if (!isAccountAllowed(invocationContext)) {
            return Future.succeededFuture(
                    InvocationResultImpl.<AuctionRequestPayload>builder()
                            .status(InvocationStatus.success)
                            .action(InvocationAction.no_action)
                            .moduleContext(invocationContext.moduleContext())
                            .build());
        }

        final ModuleContext moduleContext = addEvidenceToContext(
                (ModuleContext) invocationContext.moduleContext(),
                builder -> collectEvidence(builder, payload.bidRequest()));

        return Future.succeededFuture(
                InvocationResultImpl.<AuctionRequestPayload>builder()
                        .status(InvocationStatus.success)
                        .action(InvocationAction.update)
                        .payloadUpdate(freshPayload -> updatePayload(freshPayload, moduleContext.collectedEvidence()))
                        .moduleContext(moduleContext)
                        .build()
        );
    }

    private boolean isAccountAllowed(AuctionInvocationContext invocationContext) {
        final AccountFilter accountFilter = moduleConfig.getAccountFilter();
        final List<String> allowList = ObjectUtil.getIfNotNull(accountFilter, AccountFilter::getAllowList);
        if (CollectionUtils.isEmpty(allowList)) {
            return true;
        }
        return Optional.ofNullable(invocationContext)
                .map(AuctionInvocationContext::auctionContext)
                .map(AuctionContext::getAccount)
                .map(Account::getId)
                .filter(StringUtils::isNotBlank)
                .map(allowList::contains)
                .orElse(false);
    }

    private ModuleContext addEvidenceToContext(ModuleContext moduleContext,
                                               Consumer<CollectedEvidence.CollectedEvidenceBuilder> evidenceInjector) {

        final CollectedEvidence.CollectedEvidenceBuilder evidenceBuilder = Optional.ofNullable(moduleContext)
                .map(ModuleContext::collectedEvidence)
                .map(CollectedEvidence::toBuilder)
                .orElseGet(CollectedEvidence::builder);

        evidenceInjector.accept(evidenceBuilder);

        return Optional.ofNullable(moduleContext)
                .map(ModuleContext::toBuilder)
                .orElseGet(ModuleContext::builder)
                .collectedEvidence(evidenceBuilder.build())
                .build();
    }

    private void collectEvidence(CollectedEvidence.CollectedEvidenceBuilder evidenceBuilder, BidRequest bidRequest) {
        final Device device = bidRequest.getDevice();
        if (device == null) {
            return;
        }
        final String ua = device.getUa();
        if (ua != null) {
            evidenceBuilder.deviceUA(ua);
        }
        final UserAgent sua = device.getSua();
        if (sua != null) {
            evidenceBuilder.secureHeaders(SecureHeadersRetriever.retrieveFrom(sua));
        }
    }

    private AuctionRequestPayload updatePayload(AuctionRequestPayload existingPayload,
                                                CollectedEvidence collectedEvidence) {

        final BidRequest currentRequest = existingPayload.bidRequest();
        final BidRequest patchedRequest = enrichDevice(currentRequest, collectedEvidence);
        // todo: 'patchedRequest == currentRequest' doesn't make any sense then
        if (patchedRequest == null || patchedRequest == currentRequest) {
            return existingPayload;
        }
        return AuctionRequestPayloadImpl.of(patchedRequest);
    }

    private BidRequest enrichDevice(BidRequest bidRequest, CollectedEvidence collectedEvidence) {
        if (bidRequest == null) {
            return null;
        }
        final Device existingDevice = ObjectUtil.firstNonNull(bidRequest::getDevice, () -> Device.builder().build());

        final CollectedEvidence.CollectedEvidenceBuilder evidenceBuilder = collectedEvidence.toBuilder();
        collectEvidence(evidenceBuilder, bidRequest);

        final EnrichmentResult mergeResult = deviceEnricher.populateDeviceInfo(existingDevice, evidenceBuilder.build());
        return Optional.ofNullable(mergeResult)
                .map(EnrichmentResult::enrichedDevice)
                .map(mergedDevice -> bidRequest.toBuilder().device(mergedDevice).build())
                .orElse(null);
    }
}
```

## DeviceEnricher

```
public class DeviceEnricher {

    private static final String EXT_DEVICE_ID_KEY = "fiftyonedegrees_deviceId";

    private static final Map<String, Integer> DEVICE_FIELD_MAPPING = Map.ofEntries(
            Map.entry("Phone", OrtbDeviceType.PHONE.ordinal()),
            Map.entry("Console", OrtbDeviceType.SET_TOP_BOX.ordinal()),
            Map.entry("Desktop", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()),
            Map.entry("EReader", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()),
            Map.entry("IoT", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
            Map.entry("Kiosk", OrtbDeviceType.OOH_DEVICE.ordinal()),
            Map.entry("MediaHub", OrtbDeviceType.SET_TOP_BOX.ordinal()),
            Map.entry("Mobile", OrtbDeviceType.MOBILE_TABLET.ordinal()),
            Map.entry("Router", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
            Map.entry("SmallScreen", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
            Map.entry("SmartPhone", OrtbDeviceType.MOBILE_TABLET.ordinal()),
            Map.entry("SmartSpeaker", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
            Map.entry("SmartWatch", OrtbDeviceType.CONNECTED_DEVICE.ordinal()),
            Map.entry("Tablet", OrtbDeviceType.TABLET.ordinal()),
            Map.entry("Tv", OrtbDeviceType.CONNECTED_TV.ordinal()),
            Map.entry("Vehicle Display", OrtbDeviceType.PERSONAL_COMPUTER.ordinal()));

    private final Pipeline pipeline;

    public DeviceEnricher(Pipeline pipeline) {
        this.pipeline = Objects.requireNonNull(pipeline);
    }

    public EnrichmentResult populateDeviceInfo(Device device, CollectedEvidence collectedEvidence) {
        try (FlowData data = pipeline.createFlowData()) {
            data.addEvidence(pickRelevantFrom(collectedEvidence));
            data.process();
            final DeviceData deviceData = data.get(DeviceData.class);
            if (device == null) {
                return null;
            }
            return patchDevice(device, deviceData);
        } catch (Exception e) {
            return EnrichmentResult.builder().processingException(e).build();
        }
    }

    private Map<String, String> pickRelevantFrom(CollectedEvidence collectedEvidence) {
        final Map<String, String> evidence = new HashMap<>();

        final String ua = collectedEvidence.deviceUA();
        if (StringUtils.isNotBlank(ua)) {
            evidence.put("header.user-agent", ua);
        }
        final Map<String, String> secureHeaders = collectedEvidence.secureHeaders();
        if (MapUtils.isNotEmpty(secureHeaders)) {
            evidence.putAll(secureHeaders);
        }
        if (!evidence.isEmpty()) {
            return evidence;
        }

        Stream.ofNullable(collectedEvidence.rawHeaders())
                .flatMap(Collection::stream)
                .forEach(rawHeader -> evidence.put("header." + rawHeader.getKey(), rawHeader.getValue()));

        return evidence;
    }

    private EnrichmentResult patchDevice(Device device, DeviceData deviceData) {

        final List<String> updatedFields = new ArrayList<>();
        final Device.DeviceBuilder deviceBuilder = device.toBuilder();

        final Integer currentDeviceType = device.getDevicetype();
        if (!isPositive(currentDeviceType)) {
            final String rawDeviceType = getSafe(deviceData, DeviceData::getDeviceType);
            if (rawDeviceType != null) {
                final Integer properDeviceType = convertDeviceType(rawDeviceType);
                if (isPositive(properDeviceType)) {
                    deviceBuilder.devicetype(properDeviceType);
                    updatedFields.add("devicetype");
                }
            }
        }

        final String currentMake = device.getMake();
        if (StringUtils.isBlank(currentMake)) {
            final String make = getSafe(deviceData, DeviceData::getHardwareVendor);
            if (StringUtils.isNotBlank(make)) {
                deviceBuilder.make(make);
                updatedFields.add("make");
            }
        }

        if (StringUtils.isBlank(device.getModel())) {
            final String model = getSafe(deviceData, DeviceData::getHardwareModel);
            if (StringUtils.isNotBlank(model)) {
                deviceBuilder.model(model);
                updatedFields.add("model");
            } else {
                final List<String> names = getSafe(deviceData, DeviceData::getHardwareName);
                if (CollectionUtils.isNotEmpty(names)) {
                    deviceBuilder.model(String.join(",", names));
                    updatedFields.add("model");
                }
            }
        }

        if (StringUtils.isBlank(device.getOs())) {
            final String os = getSafe(deviceData, DeviceData::getPlatformName);
            if (StringUtils.isNotBlank(os)) {
                deviceBuilder.os(os);
                updatedFields.add("os");
            }
        }

        if (StringUtils.isBlank(device.getOsv())) {
            final String osv = getSafe(deviceData, DeviceData::getPlatformVersion);
            if (StringUtils.isNotBlank(osv)) {
                deviceBuilder.osv(osv);
                updatedFields.add("osv");
            }
        }

        if (!isPositive(device.getH())) {
            final Integer h = getSafe(deviceData, DeviceData::getScreenPixelsHeight);
            if (isPositive(h)) {
                deviceBuilder.h(h);
                updatedFields.add("h");
            }
        }

        if (!isPositive(device.getW())) {
            final Integer w = getSafe(deviceData, DeviceData::getScreenPixelsWidth);
            if (isPositive(w)) {
                deviceBuilder.w(w);
                updatedFields.add("w");
            }
        }

        if (!isPositive(device.getPpi())) {
            final Integer pixelsHeight = getSafe(deviceData, DeviceData::getScreenPixelsHeight);
            if (pixelsHeight != null) {
                final Double inchesHeight = getSafe(deviceData, DeviceData::getScreenInchesHeight);
                if (isPositive(inchesHeight)) {
                    deviceBuilder.ppi((int) Math.round(pixelsHeight / inchesHeight));
                    updatedFields.add("ppi");
                }
            }
        }

        final BigDecimal currentPixelRatio = device.getPxratio();
        if (!(currentPixelRatio != null && currentPixelRatio.intValue() > 0)) {
            final Double rawRatio = getSafe(deviceData, DeviceData::getPixelRatio);
            if (isPositive(rawRatio)) {
                deviceBuilder.pxratio(BigDecimal.valueOf(rawRatio));
                updatedFields.add("pxratio");
            }
        }

        final String currentDeviceId = getDeviceId(device);
        if (StringUtils.isBlank(currentDeviceId)) {
            final String deviceID = getSafe(deviceData, DeviceData::getDeviceId);
            if (StringUtils.isNotBlank(deviceID)) {
                setDeviceId(deviceBuilder, device, deviceID);
                updatedFields.add("ext." + EXT_DEVICE_ID_KEY);
            }
        }

        if (updatedFields.isEmpty()) {
            return null;
        }

        return EnrichmentResult.builder()
                .enrichedDevice(deviceBuilder.build())
                .enrichedFields(updatedFields)
                .build();
    }

    private static boolean isPositive(Integer value) {
        return value != null && value > 0;
    }

    private static boolean isPositive(Double value) {
        return value != null && value > 0;
    }

    /**
     * Consists of four components separated by a hyphen symbol:
     * Hardware-Platform-Browser-IsCrawler where
     * each Component represents an ID of the corresponding Profile.
     *
     * @see fiftyone.devicedetection.hash.engine.onpremise.data.DeviceDataHash#getDeviceId()
     */
    private static String getDeviceId(Device device) {

        final ExtDevice ext = device.getExt();
        if (ext == null) {
            return null;
        }
        final JsonNode savedValue = ext.getProperty(EXT_DEVICE_ID_KEY);
        return (savedValue != null && savedValue.isTextual()) ? savedValue.textValue() : null;
    }

    /**
     * Consists of four components separated by a hyphen symbol:
     * Hardware-Platform-Browser-IsCrawler where
     * each Component represents an ID of the corresponding Profile.
     *
     * @param deviceBuilder Writable builder to save device ID into.
     * @param device        Raw (non-builder) form of device before modification.
     * @param deviceId      New Device ID value.
     * @see fiftyone.devicedetection.hash.engine.onpremise.data.DeviceDataHash#getDeviceId()
     */
    private static void setDeviceId(Device.DeviceBuilder deviceBuilder, Device device, String deviceId) {

        ExtDevice ext = null;
        if (device != null) {
            ext = device.getExt();
        }
        if (ext == null) {
            ext = ExtDevice.empty();
        }
        ext.addProperty(EXT_DEVICE_ID_KEY, new TextNode(deviceId));
        deviceBuilder.ext(ext);
    }

    private <T> T getSafe(DeviceData deviceData, Function<DeviceData, AspectPropertyValue<T>> propertyGetter) {
        try {
            final AspectPropertyValue<T> propertyValue = propertyGetter.apply(deviceData);
            if (propertyValue != null && propertyValue.hasValue()) {
                return propertyValue.getValue();
            }
        } catch (Exception e) {
            // nop -- not interested in errors on getting missing values.
        }
        return null;
    }

    private Integer convertDeviceType(String deviceType) {
        return Optional.ofNullable(DEVICE_FIELD_MAPPING.get(deviceType)).orElse(OrtbDeviceType.UNKNOWN.ordinal());
    }
}

```

## PipelineBuilder
```
public class PipelineBuilder {

    private static final Collection<String> PROPERTIES_USED = List.of(
            "devicetype",
            "hardwarevendor",
            "hardwaremodel",
            "hardwarename",
            "platformname",
            "platformversion",
            "screenpixelsheight",
            "screenpixelswidth",
            "screeninchesheight",
            "pixelratio",

            "BrowserName",
            "BrowserVersion",
            "IsCrawler",

            "BrowserVendor",
            "PlatformVendor",
            "Javascript",
            "GeoLocation",
            "HardwareModelVariants");

    private PipelineBuilder() {

    }

    public static Pipeline build(ModuleConfig moduleConfig) throws Exception {
        final DataFile dataFile = moduleConfig.getDataFile();
        final DeviceDetectionOnPremisePipelineBuilder builder = makeRawBuilder(dataFile);
        applyUpdateOptions(builder, dataFile.getUpdate());
        applyPerformanceOptions(builder, moduleConfig.getPerformance());
        PROPERTIES_USED.forEach(builder::setProperty);
        return builder.build();
    }

    private static DeviceDetectionOnPremisePipelineBuilder makeRawBuilder(DataFile dataFile) throws Exception {
        final Boolean shouldMakeDataCopy = dataFile.getMakeTempCopy();
        return new DeviceDetectionPipelineBuilder()
                .useOnPremise(dataFile.getPath(), BooleanUtils.isTrue(shouldMakeDataCopy));
    }

    private static void applyUpdateOptions(DeviceDetectionOnPremisePipelineBuilder pipelineBuilder,
                                           DataFileUpdate updateConfig) {

        pipelineBuilder.setDataUpdateService(new DataUpdateServiceDefault());

        final Boolean auto = updateConfig.getAuto();
        if (auto != null) {
            pipelineBuilder.setAutoUpdate(auto);
        }

        final Boolean onStartup = updateConfig.getOnStartup();
        if (onStartup != null) {
            pipelineBuilder.setDataUpdateOnStartup(onStartup);
        }

        final String url = updateConfig.getUrl();
        if (StringUtils.isNotBlank(url)) {
            pipelineBuilder.setDataUpdateUrl(url);
        }

        final String licenseKey = updateConfig.getLicenseKey();
        if (StringUtils.isNotBlank(licenseKey)) {
            pipelineBuilder.setDataUpdateLicenseKey(licenseKey);
        }

        final Boolean watchFileSystem = updateConfig.getWatchFileSystem();
        if (watchFileSystem != null) {
            pipelineBuilder.setDataFileSystemWatcher(watchFileSystem);
        }

        final Integer pollingInterval = updateConfig.getPollingInterval();
        if (pollingInterval != null) {
            pipelineBuilder.setUpdatePollingInterval(pollingInterval);
        }
    }

    private static void applyPerformanceOptions(DeviceDetectionOnPremisePipelineBuilder pipelineBuilder,
                                                PerformanceConfig performanceConfig) {

        final String profile = performanceConfig.getProfile();
        if (StringUtils.isNotBlank(profile)) {
            for (Constants.PerformanceProfiles nextProfile : Constants.PerformanceProfiles.values()) {
                if (StringUtils.equalsAnyIgnoreCase(nextProfile.name(), profile)) {
                    pipelineBuilder.setPerformanceProfile(nextProfile);
                    // todo: return or break the cycle?
                    return;
                }
            }
        }

        final Integer concurrency = performanceConfig.getConcurrency();
        if (concurrency != null) {
            pipelineBuilder.setConcurrency(concurrency);
        }

        final Integer difference = performanceConfig.getDifference();
        if (difference != null) {
            pipelineBuilder.setDifference(difference);
        }

        final Boolean allowUnmatched = performanceConfig.getAllowUnmatched();
        if (allowUnmatched != null) {
            pipelineBuilder.setAllowUnmatched(allowUnmatched);
        }

        final Integer drift = performanceConfig.getDrift();
        if (drift != null) {
            pipelineBuilder.setDrift(drift);
        }
    }
}
```

## SecureHeaderRetriever
```
public class SecureHeadersRetriever {

    private SecureHeadersRetriever() {
    }

    public static Map<String, String> retrieveFrom(UserAgent userAgent) {

        final Map<String, String> secureHeaders = new HashMap<>();

        if (userAgent == null) {
            return secureHeaders;
        }

        final List<BrandVersion> versions = userAgent.getBrowsers();
        if (CollectionUtils.isNotEmpty(versions)) {
            final String fullUA = brandListToString(versions);
            secureHeaders.put("header.Sec-CH-UA", fullUA);
            secureHeaders.put("header.Sec-CH-UA-Full-Version-List", fullUA);
        }

        final BrandVersion platform = userAgent.getPlatform();
        if (platform != null) {
            final String platformName = platform.getBrand();
            if (StringUtils.isNotBlank(platformName)) {
                secureHeaders.put("header.Sec-CH-UA-Platform", toHeaderSafe(platformName));
            }

            final List<String> platformVersions = platform.getVersion();
            if (CollectionUtils.isNotEmpty(platformVersions)) {
                final StringBuilder stringBuilder = new StringBuilder();
                stringBuilder.append('"');
                appendVersionList(stringBuilder, platformVersions);
                stringBuilder.append('"');
                secureHeaders.put("header.Sec-CH-UA-Platform-Version", stringBuilder.toString());
            }
        }

        final Integer isMobile = userAgent.getMobile();
        if (isMobile != null) {
            secureHeaders.put("header.Sec-CH-UA-Mobile", "?" + isMobile);
        }

        final String architecture = userAgent.getArchitecture();
        if (StringUtils.isNotBlank(architecture)) {
            secureHeaders.put("header.Sec-CH-UA-Arch", toHeaderSafe(architecture));
        }

        final String bitness = userAgent.getBitness();
        if (StringUtils.isNotBlank(bitness)) {
            secureHeaders.put("header.Sec-CH-UA-Bitness", toHeaderSafe(bitness));
        }

        final String model = userAgent.getModel();
        if (StringUtils.isNotBlank(model)) {
            secureHeaders.put("header.Sec-CH-UA-Model", toHeaderSafe(model));
        }

        return secureHeaders;
    }

    private static String toHeaderSafe(String rawValue) {
        return '"' + rawValue.replace("\"", "\\\"") + '"';
    }

    private static String brandListToString(List<BrandVersion> versions) {

        final StringBuilder stringBuilder = new StringBuilder();
        for (BrandVersion nextBrandVersion : versions) {
            final String brandName = nextBrandVersion.getBrand();
            if (brandName == null) {
                continue;
            }
            if (!stringBuilder.isEmpty()) {
                stringBuilder.append(", ");
            }
            stringBuilder.append(toHeaderSafe(brandName));
            stringBuilder.append(";v=\"");
            appendVersionList(stringBuilder, nextBrandVersion.getVersion());
            stringBuilder.append('"');
        }
        return stringBuilder.toString();
    }

    private static void appendVersionList(StringBuilder stringBuilder, List<String> versions) {

        if (CollectionUtils.isEmpty(versions)) {
            return;
        }
        boolean isFirstVersionFragment = true;
        for (String nextFragment : versions) {
            if (!isFirstVersionFragment) {
                stringBuilder.append('.');
            }
            stringBuilder.append(nextFragment);
            isFirstVersionFragment = false;
        }
    }
}
```